velox/type/tz/TimeZoneMap.cpp

```diff
@@ -62,8 +62,16 @@ TTimeZoneDatabase buildTimeZoneDatabase(
     // Every single other time zone entry (outside of offsets) needs to be
     // available in external/date or this will throw.
     else {
-      timeZonePtr = std::make_unique<TimeZone>(
-          entry.second, entry.first, date::locate_zone(entry.second));
+      const date::time_zone* zone;
+      try {
+        zone = date::locate_zone(entry.second);
+      } catch (std::runtime_error& err) {
+        // Timezone not found in OS, skip it.
+        LOG(WARNING) << "Timezone [" << entry.second << "] not found due to: '"
+                     << err.what() << "', ignoring it. ";
+        continue;
+      }
```

Comment:
@pedroerp Would you think it's OK to modify `tz.h` / `tz.cpp`? Since perhaps we could add a new function, like `bool has_zone(string tz_name)` to avoid catching a `runtime_error` here.

Reply:
We made some other changes to that library, so it should be ok. Just make sure you calorie the changes and a .patch file

Reply:
Also, what happens when API clients try to convert a timezone not found?

Reply:
> Also, what happens when API clients try to convert a timezone not found?

An `Exception: VeloxUserError` like the following one will be raised

```
unknown file: Failure
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Unknown time zone: 'America/Non_Exist'
Retriable: False
Function: locateZone
File: /opt/gluten/ep/build-velox/build/velox_ep/velox/type/tz/TimeZoneMap.cpp
Line: 276
```

Reply:
@zhztheplayer there are two different things in this API:

1. The pointer to date::time_zone*, which does the actual time zone conversion.
2. The TimeZone object, which carries the date::time_zone* pointer above.

These two used to be coupled, so you could only have 2 with 1, and the TimeZone class has an internal assumption that if 1 is nullptr, it goes ahead and makes the to_sys() and to_local() based on offsets.

Now, with this PR, we are making a change where we allow 2 to be created without 1. We need to have additional logic to prevent one from making time zone conversion if the internal date::time_zone pointer doesn't exist. We still need to allow 1 to exist without 2 if we want to support missing timezone in the local tzdata database, but we need to add code to prevent time zone conversion to happen using these "empty" TimeZone objects.

Reply:
Thank you for the explanation. Which helps me understand the code's history a lot.

So, IIUC, one of the intentions of https://github.com/facebookincubator/velox/pull/10577 was to allow Velox convert time zones that don't exist in OS's timezone list, is that correct? So we'd rather to make the convertion continue than to throw an error, am I understanding correctly?

Reply:
This is the worst behaviour in Gluten before https://github.com/facebookincubator/velox/pull/10577:

1. During native library is loaded (process startup): `No error`
2. Converting an non-existing time zone: `Error`

The behaviour in Gluten after https://github.com/facebookincubator/velox/pull/10577:

1. During native library is loaded (process startup): `Error` (some zones don't exist in local tz list)
2. Converting an non-existing time zone: `N/A`

So I think we should make both 1 and 2 pass without throwing, which was the intention of https://github.com/facebookincubator/velox/pull/10577 and relevant issues, is that correct?

Reply:
> So, IIUC, one of the intentions of https://github.com/facebookincubator/velox/pull/10577 was to allow Velox convert time zones that don't exist in OS's timezone list, so we'd rather make the convertion continue than throw an error, am I understanding correctly?

No, these are different. The PR above is to support conversions using fixed timezone offsets (things like "+09:00" and "-03:00"). These also don't exist in the time zone database, but they are different from a real time zone (say, "America/Los_Angeles") which may happen not to be available in the time zone database. For the latter case, we need to fail the conversion since we don't know the time zone potential daylight savings schedule.